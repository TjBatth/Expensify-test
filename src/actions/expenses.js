
import uuid from 'uuid';
//ADDEXPENSES
export const addExpense=({
    description='',
    note='',
    amount='',
    createedAt=0

})=>({
    type:'EXPENSE_ADD',
    expense:{
    id:uuid(),
    description,
    note,
    amount,
    createedAt
    }
})

//RemoveExpenses
export const removeExpense=({id})=>({
    type:'EXPENSE_REMOVE',
    id
})

//EditExpense
export const editExpense=(id,updater)=>({
    type:'EXPENSE_EDIT',
    id,
    updater
})