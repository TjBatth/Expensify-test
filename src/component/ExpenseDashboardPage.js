import React from 'react'
import ExpesneList from './ExpenseList'
import EXpenseListFilters from './ExpenseListFilters'


const ExpenseDashboardPage=(props)=>(
    <div>
      <EXpenseListFilters />
      <ExpesneList />
      
    </div>
)




export default ExpenseDashboardPage;