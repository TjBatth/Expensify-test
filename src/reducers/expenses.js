const expensesReducersDefaultSate=[];

const expensesReducer=(state=expensesReducersDefaultSate,action)=>{ 
    switch(action.type){
        case 'EXPENSE_ADD':
        return [...state,action.expense];
        case 'EXPENSE_REMOVE':
        return state.filter((expense)=>expense.id!=action.id);
        case 'EXPENSE_EDIT':
        return state.map((expense)=>{
            console.log('mate',action.id);
          if(expense.id==action.id){ 
            return { 
                ...expense,
                ...action.updater
            }
          }else{
             return expense
           }
        })
        default:
        return state;
    }
}

export default expensesReducer;