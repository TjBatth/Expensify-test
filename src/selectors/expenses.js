import moment from 'moment';

// Get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
  //console.log('selectFilter ',expenses)
  return expenses.filter((expense) => {
    const createedAtMoment = moment(expense.createedAt);
    // const startDateMatch = startDate ? startDate.isSameOrBefore(createedAtMoment, 'day') : true;
    const startDateMatch =  true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createedAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
   // console.log("createedAtMoment "+createedAtMoment+"startDateMatch ",startDateMatch+"endDateMatch ",endDateMatch+"textMatch"+textMatch)

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createedAt < b.createedAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};