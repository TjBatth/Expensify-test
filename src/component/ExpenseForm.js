import React,{Component} from 'react'
import { sortByAmount } from '../actions/filters';
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'


 

export default class ExpenseForm extends Component{
        constructor(props){
            super(props);
            this.state={
                description:props.expense?props.expense.description:'',
                amount:props.expense?(props.expense.amount/100).toString():'',
                note:props.expense?props.expense.note:'',
                createdAt:props.expense?moment(props.expense.createdAt):moment(),
                focused:false,
                error:''
            }
        }
  
    
    changeDescription=(e)=>{
        const description=e.target.value;
        this.setState(()=>({
            description
        }))
    }
    changeAmount=(e)=>{
        const amount=e.target.value;
    if(!amount || amount.match(/^\d*(\.\d{0,2})?$/))
    {
        this.setState(()=>({
            amount
        }))
    }
    }

    changeNote=(e)=>{
        const note=e.target.value;
        this.setState(()=>({
            note
        }))
    }
    onFocusChanged=({focused})=>{
        this.setState(()=>({focused}));
    }

    createdAt=()=>{


      }

    onDateChange=(createdAt)=>{
        if(createdAt)
         this.setState(()=>({createdAt}))
    }

    addExpesneClick=()=>{
        console.log('clicked to add');
    }
    onSubmit=(e)=>{
        e.preventDefault();
          if(!this.state.description || !this.state.amount){
             this.setState(()=>({
                 error:'you should provide Description and Amount'
             }))
          }else{
            this.setState(()=>({
                error:''
            }))
            this.props.onSubmit({
                description:this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt:this.state.createdAt.valueOf(),
                note:this.state.note
            })
          }

        console.log('hello')
    }

    render(){
      return(
           <div>
           {this.state.error && <p>{this.state.error} </p>}
           <form onSubmit={this.onSubmit}>
             <input type='text'
             placeholder='descripton'
             autoFocus
             value={this.state.description}
             onChange={this.changeDescription.bind(this)}
             />
             <input type='number'
             placeholder="Amount"
             value={this.state.amount}
             onChange={this.changeAmount.bind(this)}
             />
             <textarea 
             placeholder="Note"
             value={this.state.note}
             onChange={this.changeNote.bind(this)}
             />
             <SingleDatePicker
             date={this.state.createdAt}
             onDateChange={this.onDateChange}
             focused={this.state.focused}
             onFocusChange={this.onFocusChanged}
             numberOfMonths={1}
             isOutsideRange={()=>false}
             />
             <button
                 onClick={this.addExpesneClick.bind(this)}
             >Add Expense</button> 
             </form>
           </div>
      );
    }
}