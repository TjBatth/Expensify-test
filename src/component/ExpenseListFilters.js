import React,{Component} from 'react'
import { connect } from 'react-redux'
import {setTextFilter,sortByDate,sortByAmount} from '../actions/filters'



class ExpenseListFilters extends Component {
    onChangeText(e){
       // console.log(e.target.value); 
        const value=e.target.value;
        this.props.dispatch(setTextFilter(value));   
    }
    selectSortBy=(e)=>{
        const value=e.target.value;
        if(value==='date'){
              this.props.dispatch(sortByDate());
        }else if(value === 'amount'){
            this.props.dispatch(sortByAmount());
        }
    }
    render(){
     return(
        <div>
         <input type='text'
          value={this.props.filter.text}
          onChange={this.onChangeText.bind(this)}
         />
         <select 
            value={this.props.filter.sortBy}
            onChange={this.selectSortBy.bind(this)}
        >
         <option value='date'>date</option>
         <option value='amount'>amount</option>
         </select>
        </div>
        
     );
    }
}
const stateMapToProps = (state) => ({
   filter:state.filter
})

export default connect(stateMapToProps)(ExpenseListFilters);