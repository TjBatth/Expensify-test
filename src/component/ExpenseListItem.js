import React from 'react'
import { connect } from 'react-redux'
import {Link} from 'react-router-dom'
import {removeExpense} from '../actions/expenses'

export const ExpenseListItem = ({dispatch,id,description,amount,createedAt}) =>
 {
    console.log('id',id,'des : ',description);
    return(
       <div>
       <Link to= {`/edit/${id}`}>
       <h3>{description}</h3>
       </Link>
       <p>{amount}  -  {createedAt}</p>

       <button onClick={()=>{dispatch(removeExpense({id}))}}>Remove Item</button>
       </div>
    );
}

export default connect()(ExpenseListItem);