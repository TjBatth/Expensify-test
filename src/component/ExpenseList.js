import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem';
import SelectExpense from '../selectors/expenses';

export const ExpenseList = (props) => {
  console.log('props.expenses',props.expenses)
    return(
       <div>    
          {props.expenses.map((expense) =>  
            <ExpenseListItem key={expense.id} {...expense} />)}
       </div>

    );

}

const mapStateToProps=(state)=>{
    console.log('state',state.expenses)
    return({
      expenses:SelectExpense(state.expenses,{...state.filter})
     // filter:state.filter
    })
  }

export default connect(mapStateToProps)(ExpenseList) ;