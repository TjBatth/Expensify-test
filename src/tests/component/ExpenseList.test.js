import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseList} from '../../component/ExpenseList';
import toJSON from 'enzyme-to-json'
import expenses from '../fixtures/expenses'

test('should Expense List renderer   successfully here',()=>{
      const wrapper=shallow(<ExpenseList expenses={expenses}/>)
      expect(toJSON(wrapper)).toMatchSnapshot();
})

test("Should renderer with default values : ",()=>{
    const wrapper=shallow(<ExpenseList expenses={[]}/>)
    expect(toJSON(wrapper)).toMatchSnapshot();
})