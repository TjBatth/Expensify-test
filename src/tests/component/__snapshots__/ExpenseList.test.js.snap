// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should renderer with default values :  1`] = `
<div>
  if(props.expenses.length==0)?(
  <p>
     Nothing is here to show up
  </p>
  ):(
  )
</div>
`;

exports[`should Expense List renderer   successfully here 1`] = `
<div>
  if(props.expenses.length==0)?(
  <p>
     Nothing is here to show up
  </p>
  ):(
  <Connect(ExpenseListItem)
    amount={195}
    createedAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={109500}
    createedAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={4500}
    createedAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
  )
</div>
`;
